File: lib/services/download_service.dart
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import 'package:path/path.dart' as p;
import 'youtube_service.dart';
import 'database_service.dart';
import '../models/song.dart';

class DownloadService {
  static final DownloadService _instance = DownloadService._internal();
  factory DownloadService() => _instance;
  DownloadService._internal();

  final YouTubeService _ytService = YouTubeService();
  final DatabaseService _dbService = DatabaseService();
  final Map<String, double> _downloadProgress = {};

  double getProgress(String songId) => _downloadProgress[songId] ?? 0.0;
  bool isDownloading(String songId) => _downloadProgress.containsKey(songId);

  Future<String> get _downloadDir async {
    final dir = await getApplicationDocumentsDirectory();
    final musicDir = Directory(p.join(dir.path, 'music'));
    if (!await musicDir.exists()) {
      await musicDir.create(recursive: true);
    }
    return musicDir.path;
  }

  Future<bool> downloadSong(Song song, {Function(double)? onProgress}) async {
    if (_downloadProgress.containsKey(song.id)) return false;

    try {
      _downloadProgress[song.id] = 0.0;

      final stream = await _ytService.getAudioStream(song.id);
      if (stream == null) {
        _downloadProgress.remove(song.id);
        return false;
      }

      final totalSize = await _ytService.getAudioStreamSize(song.id);
      final dir = await _downloadDir;
      final filePath = p.join(dir, '${song.id}.m4a');
      final file = File(filePath);
      final sink = file.openWrite();

      int received = 0;
      await for (final chunk in stream) {
        sink.add(chunk);
        received += chunk.length;
        if (totalSize != null && totalSize > 0) {
          final progress = received / totalSize;
          _downloadProgress[song.id] = progress;
          onProgress?.call(progress);
        }
      }

      await sink.close();
      _downloadProgress.remove(song.id);

      await _dbService.updateSongDownload(song.id, true, filePath);
      return true;
    } catch (e) {
      _downloadProgress.remove(song.id);
      print('Download error: $e');
      return false;
    }
  }

  Future<bool> deleteSong(String songId) async {
    try {
      final song = await _dbService.getSong(songId);
      if (song?.filePath != null) {
        final file = File(song!.filePath!);
        if (await file.exists()) {
          await file.delete();
        }
      }
      await _dbService.updateSongDownload(songId, false, null);
      return true;
    } catch (e) {
      print('Delete error: $e');
      return false;
    }
  }

  Future<String?> getLocalPath(String songId) async {
    final song = await _dbService.getSong(songId);
    if (song?.isDownloaded == true && song?.filePath != null) {
      final file = File(song!.filePath!);
      if (await file.exists()) {
        return song.filePath;
      }
    }
    return null;
  }
}

File: react-music-app/src/services/DownloadService.ts
import * as FileSystem from 'expo-file-system';
import DatabaseService from './DatabaseService';
import YouTubeService from './YouTubeService';
import { Song } from '../models/Song';

const MUSIC_DIR = `${FileSystem.documentDirectory}music/`;

class DownloadService {
  private static instance: DownloadService;
  private _progress: Map<string, number> = new Map();
  private _activeDownloads: Set<string> = new Set();

  static getInstance(): DownloadService {
    if (!DownloadService.instance) {
      DownloadService.instance = new DownloadService();
    }
    return DownloadService.instance;
  }

  async ensureMusicDir(): Promise<void> {
    const info = await FileSystem.getInfoAsync(MUSIC_DIR);
    if (!info.exists) {
      await FileSystem.makeDirectoryAsync(MUSIC_DIR, { intermediates: true });
    }
  }

  getLocalPath(songId: string): string {
    return `${MUSIC_DIR}${songId}.m4a`;
  }

  async isDownloaded(songId: string): Promise<boolean> {
    const path = this.getLocalPath(songId);
    const info = await FileSystem.getInfoAsync(path);
    return info.exists;
  }

  getProgress(songId: string): number {
    return this._progress.get(songId) ?? 0;
  }

  isDownloading(songId: string): boolean {
    return this._activeDownloads.has(songId);
  }

  async downloadSong(
    song: Song,
    onProgress?: (progress: number) => void
  ): Promise<boolean> {
    if (this._activeDownloads.has(song.id)) return false;

    try {
      await this.ensureMusicDir();
      this._activeDownloads.add(song.id);
      this._progress.set(song.id, 0);

      const streamData = await YouTubeService.getStreamData(song.id);
      if (!streamData?.url) throw new Error('No stream URL');

      const destPath = this.getLocalPath(song.id);
      const downloadResumable = FileSystem.createDownloadResumable(
        streamData.url,
        destPath,
        {},
        (downloadProgress) => {
          const progress =
            downloadProgress.totalBytesWritten /
            (downloadProgress.totalBytesExpectedToWrite || 1);
          this._progress.set(song.id, progress);
          onProgress?.(progress);
        }
      );

      const result = await downloadResumable.downloadAsync();
      if (!result?.uri) throw new Error('Download failed');

      await DatabaseService.updateSongDownload(song.id, true, destPath);
      this._progress.set(song.id, 1);
      onProgress?.(1);
      return true;
    } catch (err) {
      console.error('DownloadService.downloadSong error:', err);
      this._progress.delete(song.id);
      return false;
    } finally {
      this._activeDownloads.delete(song.id);
    }
  }

  async deleteSong(songId: string): Promise<boolean> {
    try {
      const path = this.getLocalPath(songId);
      const info = await FileSystem.getInfoAsync(path);
      if (info.exists) {
        await FileSystem.deleteAsync(path);
      }
      await DatabaseService.updateSongDownload(songId, false, undefined);
      return true;
    } catch (err) {
      console.error('DownloadService.deleteSong error:', err);
      return false;
    }
  }
}

export default DownloadService.getInstance();
